spock-specs/src/main/groovy/org/spockframework/specs/extension/SnapshotExtension.groovy

```diff
@@ -17,24 +21,63 @@ class SnapshotExtension implements IAnnotationDrivenExtension<Snapshot> {
   }
 
   @Override
-  void visitFieldAnnotation(Snapshot annotation, FieldInfo field) {
-    Assert.that(field.type.isAssignableFrom(Snapshotter), "Field must be of type Snapshotter")
-    field.parent.bottomSpec.addSetupInterceptor(new SnapshotInterceptor(annotation, field))
+  void visitSpecAnnotation(Snapshot annotation, SpecInfo specInfo) {
+    specInfo.features.each { visitFeatureAnnotation(annotation, it) }
+  }
+
+  @Override
+  void visitFeatureAnnotation(Snapshot annotation, FeatureInfo feature) {
+    def snapshotterFields = feature.spec.allFields.findAll {
+      ((it.type == Snapshotter) || (it.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !it.static && !it.shared
+    }
+    def snapshotterParameters = feature.featureMethod.parameters.findAll {
+      ((it.reflection.type == Snapshotter) || (it.reflection.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !(it.name in feature.dataVariables)
+    }
+    Assert.that(
+      (snapshotterFields.size() != 0) || (snapshotterParameters.size() != 0),
+      'There must be at least one Snapshotter parameter that is not a data variable or one Snapshotter field that is neither static nor @Shared'
+    )
+
+    feature.addIterationInterceptor(new SnapshotFieldInterceptor(annotation, snapshotterFields))
+    feature.featureMethod.addInterceptor(new SnapshotParameterInterceptor(annotation, snapshotterParameters))
+  }
+
+  @CompileStatic
+  class SnapshotFieldInterceptor implements IMethodInterceptor {
+    private final List<FieldInfo> fields
+    private final Snapshot annotation
+
+    SnapshotFieldInterceptor(Snapshot annotation, List<FieldInfo> fields) {
+      this.fields = fields
+      this.annotation = annotation
+    }
+
+    @Override
+    void intercept(IMethodInvocation invocation) throws Throwable {
+      fields
+        .findAll { it.readValue(invocation.instance) == null }
```

Comment:
Shouldn't this be an error?

Reply:
That's a matter of opinion I'd say.
When I write injecting extensions, I usually leave alone fields or parameters that are already set (i.e. fields that are not `null`, parameters that are not `MISSING_ARGUMENT`).
This way a user can initialize the field or parameter differently if he wants to.

But of course the extension could also claim ownership and fail or overwrite the value if you like that better.

Reply:
I don't think it makes sense to be lenient for this particular extension.